### PR TITLE
In game hypo

### LINKED ISSUE #1156: in-game hypothetical

### PR DESCRIPTION
**Pull request by Zamiell**
Resolves #1156

### CODE REVIEW

**Comment by Zamiell** [comment]
opening this now because i want to see the diff of the whole branch

File: client/src/game/ui/drawReplayArea.ts [hunk 1]
```diff
@@ -220,18 +220,22 @@ export default function drawReplayArea(winW: number, winH: number): void {
...
+  // The "Exit Replay" button
   globals.elements.replayExitButton = new Button({
-    x:
-      (replayButtonValues.x +
-        replayButtonValues.w +
-        replayButtonValues.spacing / 2) *
-      winW,
+    x: bottomLeftReplayButtonValues.x * winW,
     y: bottomButtonValues.y * winH,
-    width: (replayButtonValues.w * 2 + replayButtonValues.spacing * 2) * winW,
```
Comment (timotree3) [inline client/src/game/ui/drawReplayArea.ts:234]: I didn't change this on purpose

Comment (timotree3) [reply client/src/game/ui/drawReplayArea.ts:234]: oh I see that was you.

Comment (Zamiell) [reply client/src/game/ui/drawReplayArea.ts:234]: ?

**Comment by timotree3** [comment]
The feature seems to be working smoothly.

TODO:
- [ ] prompt user for card identity when playing/discarding an unknown card
- [ ] allow clues to self when cards are morphed and would be touched
- [ ] check if you rewind start a hypo and draw a card that is fully identified since you see it in the present everything still works
- [ ] desirable? - add back Show Drawn Cards button or change default to show drawn cards for the above-mentioned case (removed because it conflicts with Your Turn indicator)
- [ ] automatically morph all cards in hand according to identity notes when starting hypothetical for ease of use.
- [ ] test that shared replay still works
- [ ] fix enter hypothetical button in private post-game replay

**Comment by Zamiell** [comment]
> add back Show Drawn Cards button

how could that be possible because you dont know what drawn cards will be

**Comment by Zamiell** [comment]
the rest sound good

**Comment by timotree3** [comment]
it only shows cards that have already been drawn in the present

**Comment by Zamiell** [comment]
oh

thats a bit confusing, because 99% of the time people will be initiating a hypo from the present turn
thus, i dont think we need to add additional UI elements to facilitate initiating a hypo from a previous turn

**Comment by timotree3** [comment]
> 99% of the time people will be initiating a hypo from the present turn
maybe you would start from a past turn to see if anybody would've acted different had your held card been X?

**Comment by Zamiell** [comment]
that's true

**Comment by Zamiell** [comment]
ok, so show the button, but only if the hypo start turn is not the present turn

**Comment by timotree3** [comment]
sounds good. although it's also possible for the bottom card in the deck to be fully identified if your hand is locked

**Comment by Zamiell** [comment]
true but we probably don't care about that

File: client/src/game/reducers/initialStates/initialState.ts [hunk 2]
```diff
@@ -25,7 +38,10 @@ export default function initialState(metadata: GameMetadata): State {
     datetimeFinished: null,
 
     metadata,
-    cardIdentities: [],
+    cardIdentities: initArray(startingDeckSize, {
+      suitIndex: null,
+      rank: null,
+    }),
```
Comment (timotree3) [inline client/src/game/reducers/initialStates/initialState.ts:44]: 1. `startingDeckSize` returns the size of the deck on turn 1. Do you mean to use the size of the deck before cards are dealt?

2. Now everywhere where we checked for undefined before we need to check for {null, null}, I can update that.

3. If this is actually a good idea then there should be similar code in `loadGameJSON.ts`

Comment (Zamiell) [reply client/src/game/reducers/initialStates/initialState.ts:44]: 1) yes
2) yes
3) yes

Comment (timotree3) [reply client/src/game/reducers/initialStates/initialState.ts]: fixed

**Comment by timotree3** [comment]
I recently learned that you can be in a shared replay and a game at the same time. Are there gonna be any bugs where we expect `shared` to equal `null` during in-game hypo but it actually doesn't because the user is doing both at once?

**Comment by timotree3** [comment]
Also this PR is ready for review and merge

**Comment by Zamiell** [comment]
what do you mean "at the same time"

**Comment by Zamiell** [comment]
you can be on the server, but not the client

**Comment by Zamiell** [comment]
fantastic work, thank you

**Pull request merged**